pkg/build/controller/buildpod/factory.go

```diff
@@ -0,0 +1,213 @@
+		select {
+		case <-time.After(storeSyncedPollPeriod):
+		case <-stopCh:
+			return
+		}
+	}
+	close(ready)
+}
+
+func (bc *BuildPodController) worker() {
+	for {
+		if quit := bc.work(); quit {
+			return
+		}
+	}
+}
+
+func (bc *BuildPodController) work() bool {
+	key, quit := bc.queue.Get()
+	if quit {
+		return true
+	}
+	defer bc.queue.Done(key)
+
+	pod, err := bc.getPodByKey(key.(string))
+	if err != nil {
+		glog.Error(err.Error())
+	}
+
+	if pod == nil {
+		return false
+	}
+
+	err = bc.HandlePod(pod)
+	bc.handleErr(err, key, pod)
+
+	return false
+}
+
+func (bc *BuildPodController) getPodByKey(key string) (*kapi.Pod, error) {
+	obj, exists, err := bc.podStore.Indexer.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve pod %q from store: %v", key, err)
+		bc.queue.Add(key)
+		return nil, err
+	}
+	if !exists {
+		glog.Infof("Pod %q has been deleted", key)
+		return nil, nil
+	}
+
+	return obj.(*kapi.Pod), nil
+}
+
+func (bc *BuildPodController) updatePod(old, cur interface{}) {
+	// A periodic relist will send update events for all known pods.
+	curPod := cur.(*kapi.Pod)
+	oldPod := old.(*kapi.Pod)
+	if curPod.ResourceVersion == oldPod.ResourceVersion {
+		return
+	}
+	if isBuildPod(curPod) {
+		bc.enqueuePod(curPod)
+	}
+}
+
+func (bc *BuildPodController) deletePod(obj interface{}) {
+	pod, ok := obj.(*kapi.Pod)
+	if !ok {
+		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			glog.Errorf("Couldn't get object from tombstone: %+v", obj)
+			return
+		}
+		pod, ok = tombstone.Obj.(*kapi.Pod)
+		if !ok {
+			glog.Errorf("Tombstone contained object that is not a pod: %+v", obj)
+			return
+		}
+	}
+	if isBuildPod(pod) {
+		bc.HandleBuildPodDeletion(pod)
+	}
+}
+
+func (c *BuildPodController) enqueuePod(pod *kapi.Pod) {
+	key, err := kcontroller.KeyFunc(pod)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %#v: %v", pod, err)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (bc *BuildPodController) handleErr(err error, key interface{}, pod *kapi.Pod) {
+	if err == nil {
+		bc.queue.Forget(key)
+		return
+	}
+
+	if strategy.IsFatal(err) {
+		glog.V(3).Infof("Will not retry fatal error for pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		bc.queue.Forget(key)
+		return
+	}
+
+	if bc.queue.NumRequeues(key) < maxRetries {
+		glog.V(4).Infof("Retrying pod %s/%s: %v", pod.Namespace, pod.Name, err)
+		bc.queue.AddRateLimited(key)
+		return
+	}
+
+	glog.V(3).Infof("Giving up retrying pod %s/%s: %v", pod.Namespace, pod.Name, err)
+	bc.queue.Forget(key)
+}
+
+func isBuildPod(pod *kapi.Pod) bool {
+	return len(buildutil.GetBuildName(pod)) > 0
+}
+
+func (bc *BuildPodController) getBuildForPod(pod *kapi.Pod) (*buildapi.Build, bool, error) {
+	if !isBuildPod(pod) {
+		return nil, false, errors.New("cannot get build for pod: pod is not a build pod")
+	}
+	build, err := bc.buildStore.Builds(pod.Namespace).Get(buildutil.GetBuildName(pod))
```

Comment:
> Yes we will because we will handle the pod deleted event

?

handling the pod deleted event properly requires that we are able to find the build associated w/ the pod that we got a delete event for.  If we can't find the build because it's not in the cache yet(and we don't check the api server directly), we can't handle the event properly, because handling the event properly requires updating the build object.

again, i think you've convinced me it's not possible for the build not to be in the cache in this particular case, but the general class of problem is real.